azure-auth-helper/src/main/java/com/microsoft/azure/auth/LocalAuthServer.java

```diff
@@ -0,0 +1,149 @@
+                    throws IOException, ServletException {
+                initHtmlTemplate();
+                try {
+                    ((Request) request).setHandled(true);
+                    code = request.getParameter(Constants.CODE);
+                    // handle the error response described at
+                    // https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-protocols-oauth-code#error-response
+                    errorDescription = request.getParameter(Constants.ERROR_DESCRIPTION);
+                    error = request.getParameter(Constants.ERROR);
+
+                    final boolean isSuccess = StringUtils.isEmpty(error) && StringUtils.isNotEmpty(code);
+                    response.setStatus(HttpServletResponse.SC_OK);
+                    response.setContentType(Constants.CONTENT_TYPE_TEXT_HTML);
+                    try (final PrintWriter writer = response.getWriter()) {
+                        writer.write(isSuccess ? loginSuccessBodyTemplate : String.format(loginErrorBodyTemplate, error, errorDescription));
+                        writer.flush();
+                    }
+                    response.flushBuffer();
+                } finally {
+                    semaphore.release();
+                }
+            }
+
+        });
+    }
+
+    public URI getURI() {
+        return jettyServer.getURI();
+    }
+
+    public void start() throws IOException {
+        if (jettyServer.isStarted()) {
+            return;
+        }
+        try {
+            jettyServer.start();
+        } catch (Exception e) { // server.start will throw Exception
+            if (e instanceof RuntimeException) {
+                // avoid unnecessary exception convert
+                throw (RuntimeException) e;
+            }
+
+            throw new IOException(e);
+        }
+
+    }
+
+    public void stop() throws IOException {
+        semaphore.release();
+        try {
+            jettyServer.stop();
+        } catch (Exception e) { // server.stop will throw Exception
+            if (e instanceof RuntimeException) {
+                // avoid unnecessary exception convert
+                throw (RuntimeException) e;
+            }
+
+            throw new IOException(e);
+        }
+
+    }
+
+    public String waitForCode() throws InterruptedException, AzureLoginFailureException {
+        if (!semaphore.tryAcquire(Constants.OAUTH_TIMEOUT_MINUTES, TimeUnit.MINUTES)) {
+            try {
+                stop();
+            } catch (IOException e) {
+                // ignore
+            }
+        }
+        if (StringUtils.isEmpty(error) && StringUtils.isNotEmpty(code)) {
+            return code;
+        }
+        if (StringUtils.isNotEmpty(error)) {
+            if (StringUtils.isNotEmpty(errorDescription)) {
+                throw new AzureLoginFailureException(error + "\nReason: " + errorDescription);
+            }
+            throw new AzureLoginFailureException(error);
+        } else {
```

Comment:
You already threw in the `if` branch, is this `else` branch needed?

Reply:
add a case of timeout after 5 minutes, and move the last throw statement  to the outside of else clause.